appium-provider/src/main/java/com/paypal/selion/appium/platform/grid/SeLionAppiumAndroidDriver.java

```diff
@@ -245,27 +246,40 @@ public void clearText(WebElement webElement) {
     @Override
     public void setText(WebElement webElement, String text) {
         logger.entering(webElement);
-        //As per the UI Object API doc a text field will be cleared before setting value
+        // As per the UI Object API doc a text field will be cleared before setting value
         webElement.clear();
         webElement.sendKeys(text);
         logger.exiting();
     }
 
     @Override
     public void swipe(int startx, int starty, int endx, int endy) {
-        //super.swipe(startx, starty, endx, endy, OPERATION_DURATION_MILLI_SECONDS );
+        swipe(startx, starty, endx, endy, OPERATION_DURATION_MILLI_SECONDS);
+    }
 
-        // On Appium Android we mimic swipe via one finger tap
-        this.tap(1, starty, endx, OPERATION_DURATION_MILLI_SECONDS);
+    public void swipe(int startx, int starty, int endx, int endy, int duration) {
+        new TouchAction(this).press(startx, starty).waitAction(Duration.ofMillis(duration)).moveTo(endx, endy).release().perform();
     }
 
-    @Override
-    public void rotate(DeviceRotation rotation) {
-        // TODO
+    public void tap(int fingers, WebElement element, int duration) {
+        MultiTouchAction multiTouch = new MultiTouchAction(this);
+
+        for (int i = 0; i < fingers; i++) {
+            TouchAction tap = new TouchAction(this);
+            multiTouch.add(tap.press(element).waitAction(Duration.ofMillis(duration)).release());
+        }
+
+        multiTouch.perform();
     }
 
-    @Override
-    public DeviceRotation rotation() {
-        return null;
+    public void tap(int fingers, int x, int y, int duration) {
+        MultiTouchAction multiTouch = new MultiTouchAction(this);
+
+        for (int i = 0; i < fingers; i++) {
+            TouchAction tap = new TouchAction(this);
+            multiTouch.add(tap.press(x, y).waitAction(Duration.ofMillis(duration)).release());
+        }
+        multiTouch.perform();
     }
```

Comment:
Can you please confirm if it is multi finger tap or multi tap action or combination of both?
Based on my observation it is behaving like multi finger multi tap action.

Above implementation completely depends on the duration parameter for example if I want to perform a double tap with duration 200ms, this will become two long press actions, but if I perform a double tap action with 30ms then that will be treated as double tap. 

My suggestion would be not use raw actions rather use tap action exposed by TouchAction api-

Ex- 
```
MultiTouchAction multiTouch = new MultiTouchAction(driver);
        for (int i = 0; i < fingers; i++) {
            TouchAction action = new TouchAction(driver);
            multiTouch.add(action.tap(element));
        }
        multiTouch.perform();
```

Reply:
Observation : Most of the methods are expecting WebElement as an argument, so in test first you have to grab the WebElement instance and then only you can invoke these methods. Why can't we follow the same approach we have taken for HTML elements where we are invoking the findElement internally to grab the element reference. I am not sure if that would violate the design of the framework.